crates/data-generator/src/blocks/output.rs

```diff
@@ -0,0 +1,628 @@
+        fn from(block_state: &State) -> Self {
+            let properties = block_state
+                .properties
+                .iter()
+                .map(|block_prop| BlockStateProperty {
+                    field_name: block_prop.field.to_ident(),
+                    enum_name: block_prop.prop_enum.to_ident(),
+                    var_name: block_prop.variant.to_ident(),
+                })
+                .collect();
+
+            BlockStateInst { properties }
+        }
+    }
+
+    impl ToTokens for BlockStateInst {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            // "Ladder"
+            tokens.append("Block".to_ident());
+
+            if !self.properties.is_empty() {
+                let mut props = TokenStream::new();
+                // "facing: Facing::North, waterlogged: Waterlogged::True,"
+                self.properties
+                    .iter()
+                    .for_each(|prop| prop.to_tokens(&mut props));
+
+                // "LargeAmethystBud { facing: Facing::North, waterlogged: Waterlogged::True, }"
+                tokens.append(Group::new(Delimiter::Brace, props))
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct BlockStateProperty {
+        pub field_name: Ident, // instrument
+        pub enum_name: Ident,  // NoteBlock
+        pub var_name: Ident,   // IronXylophone
+    }
+
+    impl ToTokens for BlockStateProperty {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let field_name = &self.field_name;
+            let enum_name = &self.enum_name;
+            let var_name = &self.var_name;
+
+            // "instrument: NoteBlock::IronXylophone,"
+            tokens.extend(quote!(#field_name: #enum_name::#var_name,))
+        }
+    }
+
+    impl ToTokens for Id {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            tokens.append(Literal::u32_unsuffixed(self.0))
+        }
+    }
+}
+
+mod prop_enum {
+    use proc_macro2::{Delimiter, Group, Literal, Punct, Spacing, TokenStream};
+    use quote::{quote, ToTokens, TokenStreamExt as _};
+
+    use super::StringExt as _;
+    use crate::blocks::model::BlockProperty;
+
+    pub struct PropertyEnum {
+        pub prop_name: String, // Powered
+        pub prop_vars: Vec<PropertyEnumVariant>,
+        pub ordered: bool,
+    }
+
+    impl From<&BlockProperty> for PropertyEnum {
+        fn from(block_prop: &BlockProperty) -> Self {
+            let prop_name_pc = block_prop.name_pc.clone();
+            let ordered = block_prop.is_num();
+
+            let prop_variants: Vec<PropertyEnumVariant> = block_prop
+                .variants
+                .iter()
+                .enumerate()
+                .map(|(var_ord, var_name)| PropertyEnumVariant {
+                    var_name: var_name.defused_name().to_string(),
+                    var_ordinal: Some(block_prop)
+                        .filter(|&prop| prop.is_num())
+                        .map(|_| var_ord),
```

Comment:
This is weird way to introduce an `Option`.

Why not just:

```suggestion
                    var_ordinal: if block_prop.is_num() {
                        Some(var_ord)
                    } else {
                        None
                    },
```

Why is much easier to understand.

Even more, this is such a common use case that there is [`bool::then_some`](https://doc.rust-lang.org/std/primitive.bool.html#method.then_some):

```suggestion
                    var_ordinal: prop.is_num().then_some(var_ord),
```

Reply:
Obsolete for this change as of 4115128d4ceb2f4c64f815c965bd5728e6fff786; still, updated the rest of the code base to use `Option::then_some`.